src/System.Private.CoreLib/shared/System/Marvin.cs

```diff
@@ -28,73 +28,194 @@ internal static partial class Marvin
-                    byteOffset += 4;
-                    Block(ref p0, ref p1);
-                    goto case 3;
-
-                case 3:
-                    p0 += 0x80000000u | (((uint)(Unsafe.AddByteOffset(ref data, byteOffset + 2))) << 16) | (uint)(Unsafe.ReadUnaligned<ushort>(ref Unsafe.AddByteOffset(ref data, byteOffset)));
-                    break;
-
-                default:
-                    Debug.Fail("Should not get here.");
-                    break;
+                partialResult >>= 8; // make some room for the 0x80 byte
+                partialResult |= 0x80000000u; // put the 0x80 byte at the beginning
+                partialResult >>= (~count << 3) & 0x1F; // shift out all previously consumed bytes
             }
+            else
+            {
+                partialResult <<= 8; // make some room for the 0x80 byte
+                partialResult |= 0x80u; // put the 0x80 byte at the end
+                partialResult <<= (~count << 3) & 0x1F; // shift out all previously consumed bytes
+            }
+
+        DoFinalRoundsAndReturn:
 
+            // Now that we've computed the final partial result, merge it in and run two rounds of
+            // the block function to finish out the Marvin algorithm.
+
+            p0 += partialResult;
             Block(ref p0, ref p1);
             Block(ref p0, ref p1);
 
             return (int)(p1 ^ p0);
+
+        InputTooSmallToEnterMainLoop:
+
+            // We can't run the main loop, but we might still have 4 or more bytes available to us.
+            // If so, jump to the 4 .. 7 bytes logic immediately after the main loop.
+
+            if ((count & 4) != 0)
+            {
+                goto Between4And7BytesRemain;
+            }
+
+            // We had only 0 .. 3 bytes to begin with, so we can't perform any 32-bit reads.
+            // This means that we're going to be building up the final result right away and
+            // will only ever run two rounds total of the block function. Let's initialize
+            // the partial result to "no data".
+
+            if (BitConverter.IsLittleEndian)
+            {
+                partialResult = 0x80u;
+            }
+            else
+            {
+                partialResult = 0x80000000u;
+            }
+
+            if ((count & 1) != 0)
+            {
+                // If the buffer is 1 or 3 bytes in length, let's read a single byte now
+                // and merge it into our partial result. This will result in partialResult
+                // having one of the two values below, where AA BB CC are the buffer bytes.
+                // 
+                //                  (little-endian / big-endian)
+                // [ AA          ]  -> 0x000080AA / 0xAA800000;
+                // [ AA BB CC    ]  -> 0x000080CC / 0xCC800000;
+
+                partialResult = Unsafe.AddByteOffset(ref data, (nuint)(uint)count & 2);
```

Comment:
Is it correct that we ignore the 0x80u that was already written to partialResult?

Reply:
Yes, it's meant to be overwritten in this case. The alternative was to initialize it to zero and to write "if (evenLength) { set to 0x80 } else { read and add 0x8000 }", but perf testing showed that a single throwaway mov instruction was cheaper than the more complicated branching / jumping logic.

Reply:
Ok, thanks.